pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,310 @@
+		// so we make sure the client inside the guest cannot send another request before
+		// the previous request was processed
+		go func() {
+			if err := s.rateLimiter.Wait(ctx); err != nil {
+				return // Context canceled, just return
+			}
+
+			// This will block until a request is received or `conn` is closed by the
+			// parent goroutine or QEMU
+			requestLine, err := waitForRequest(reader)
+			if err != nil && (errors.Is(err, io.EOF) || errors.Is(err, net.ErrClosed)) {
+				// `net.ErrClosed`: The parent goroutine closed the connection
+				// `io.EOF`: QEMU closes the connection
+				close(newRequest) // this is really only required if QEMU closes the connection
+				return
+			}
+
+			// Non-blocking send, because `newRequest` is a buffered channel
+			newRequest <- reqResult{request: requestLine, err: err}
+		}()
+
+		select {
+		case res, ok := <-newRequest:
+			if !ok {
+				return // QEMU closed the connection
+			}
+
+			if res.err != nil {
+				log.Log.Reason(res.err).Warning("virtio-serial socket read failed")
+				// Let's clean the remaining data in the connection to prevent the following
+				// requests to fail due to left over bytes from the current invalid request
+				reader.Reset(conn)
+				replyError(conn)
+				continue
+			}
+
+			response, err := s.handleRequest(res.request)
+			if err != nil {
+				log.Log.Reason(err).Error("failed to process the request")
+				replyError(conn)
+				continue
+			}
+
+			err = reply(conn, response)
+			if err != nil {
+				log.Log.Reason(err).Error("failed to send the metrics")
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *downwardMetricsServer) handleRequest(requestLine string) ([]byte, error) {
+	err := parseRequest(requestLine)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := s.getXmlMetrics()
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func (s *downwardMetricsServer) getXmlMetrics() ([]byte, error) {
+	var xmlMetrics []byte
+	metrics, err := s.reportFn()
+	if err != nil {
+		xmlMetrics = []byte(emptyMetrics)
+		log.Log.Reason(err).Error("failed to collect the metrics")
+	} else {
+		xmlMetrics, err = xml.MarshalIndent(metrics, "", "  ")
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode metrics: %v", err)
+		}
+	}
+
+	// `vm-dump-metrics` expects `\n\n` as a termination symbol
+	xmlMetrics = append(xmlMetrics, '\n', '\n')
+	return xmlMetrics, nil
+}
+
+func connect(ctx context.Context, socketPath string, attempts uint) (net.Conn, error) {
+	var conn net.Conn
+	var err error
+
+	multiplier := 1
+	for i := uint(0); i < attempts; i++ {
+		conn, err = net.Dial("unix", socketPath)
+		if err == nil {
+			break
+		}
+
+		// It is only tried again in case the socket doesn't exist or not one is
+		// listening on the other end yet
+		if !(errors.Is(err, syscall.ECONNREFUSED) || errors.Is(err, syscall.ENOENT)) {
+			break
+		}
+
+		if i == attempts {
+			break // it was the last attempt
```

Comment:
nit: Returning the error here with wrapped "Reached max attempts" could be helpful in debugging.

Reply:
done